contracts/extensions/nft/NFT.sol

```diff
@@ -0,0 +1,175 @@
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+ */
+
+contract NFTExtension is DaoConstants, AdapterGuard, IExtension, IERC721Receiver {
+    using Address for address payable;
+    // Add the library methods
+    using EnumerableSet for EnumerableSet.UintSet;
+    using EnumerableSet for EnumerableSet.AddressSet;
+
+    bool public initialized = false; // internally tracks deployment under eip-1167 proxy pattern
+    DaoRegistry public dao;
+
+    enum AclFlag {COLLECT, TRANSFER, RETURN, REGISTER_NEW_NFT}
+
+    // Keeps track of the nft owners managed by the contract
+    mapping(address => mapping(address => EnumerableSet.UintSet))
+        private _tokenOwners;
+
+    address[] public nftCollection;
+    mapping(address => bool) public availableNFTs;
+
+    modifier hasExtensionAccess(IExtension extension, AclFlag flag) {
+        require(
+            dao.state() == DaoRegistry.DaoState.CREATION ||
+                dao.hasAdapterAccessToExtension(
+                    msg.sender,
+                    address(extension),
+                    uint8(flag)
+                ),
+            "nft-extension::accessDenied"
+        );
+        _;
+    }
+
+    /// @notice Clonable contract must have an empty constructor
+    constructor() {}
+
+    function initialize(DaoRegistry _dao, address creator) external override {
+        require(!initialized, "already initialized");
+        require(_dao.isActiveMember(creator), "not active member");
+
+        initialized = true;
+        dao = _dao;
+    }
+
+    /**
+     * @notice Collects the NFT from the owner and moves to the contract address
+     * @notice It must be have been allowed to move this token by either {approve} or {setApprovalForAll}.
+     */
+    function collect(
+        address owner,
+        address nftAddr,
+        uint256 nftTokenId
+    ) external {
+        require(isNFTAllowed(nftAddr), "nft not allowed");
+        IERC721 erc712 = IERC721(nftAddr);
+        // Move the NFT to the contract address
+        erc712.safeTransferFrom(owner, address(this), nftTokenId);
+        // Save the old owner, nft and tokenId to be able to return it back later on
+        _tokenOwners[owner][nftAddr].add(nftTokenId);
+    }
+
+    /**
+     * @notice Internally transfer the NFT token from the old owner to the new owner address
+     * @notice It also updates the internal state to keep track of the new other and holdings
+     * @notice It must be have been allowed to move this token by either {approve} or {setApprovalForAll}.
+     */
+    function transferFrom(
+        address owner,
+        address nftAddr,
+        uint256 nftTokenId
+    ) public hasExtensionAccess(this, AclFlag.TRANSFER) {
+        require(isNFTAllowed(nftAddr), "non allowed nft");
+        // Move the NFT to the contract address
+        IERC721 erc712 = IERC721(nftAddr);
```

Comment:
I actually like this idea of managing transfers from the extension directly. Something to think about for the Bank as well. This way we do it securely and we won't risk make a mistake in the future.

Reply:
Cool. Sounds good!